File: app-builder/frontend/src/shared/schema/blockAppearance.ts
import type { CSSProperties } from 'react'

export type BlockAppearanceProps = {
  backgroundColor?: string
  borderColor?: string
  borderWidth?: number
  borderRadius?: number
  contentPadding?: number
  textSurfaceEnabled?: boolean
}

export function appearanceNumber(value: unknown, fallback: number, max = 96): number {
  return typeof value === 'number' && Number.isFinite(value) ? Math.max(0, Math.min(max, value)) : fallback
}

export function appearanceColor(value: unknown, fallback: string): string {
  return typeof value === 'string' && /^(transparent|#[0-9a-f]{6})$/i.test(value) ? value : fallback
}

export function blockSurfaceStyle(type: string, props: Record<string, unknown>, scale = 1): CSSProperties {
  if (type === 'text' && props.textSurfaceEnabled !== true) return {}
  const width = appearanceNumber(props.borderWidth, type === 'text' ? 1 : 0, 12) * scale
  return {
    backgroundColor: appearanceColor(props.backgroundColor, type === 'button' ? '#2563eb' : type === 'text' ? '#ffffff' : 'transparent'),
    border: `${width}px solid ${appearanceColor(props.borderColor, '#cbd5e1')}`,
    borderRadius: appearanceNumber(props.borderRadius, type === 'button' ? 10 : type === 'text' ? 12 : 0, 999) * scale,
    boxSizing: 'border-box',
  }
}

export function blockPadding(type: string, props: Record<string, unknown>, scale = 1): number {
  return appearanceNumber(props.contentPadding, type === 'hero' ? 16 : 12) * scale
}

File: app-builder/native-preview/Android/app/src/main/java/com/apptura/nativepreview/renderers/BlockAppearance.kt
package com.apptura.nativepreview.renderers

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import com.apptura.nativepreview.models.Block
import kotlinx.serialization.json.JsonPrimitive

fun appearanceNumber(block: Block, key: String, fallback: Float, max: Float = 96f): Float =
    (block.props[key] as? JsonPrimitive)?.content?.toFloatOrNull()
        ?.takeIf { it.isFinite() }?.coerceIn(0f, max) ?: fallback

fun appearanceColor(block: Block, key: String, fallback: Color): Color {
    val raw = (block.props[key] as? JsonPrimitive)?.content ?: return fallback
    if (raw.equals("transparent", ignoreCase = true)) return Color.Transparent
    if (!Regex("^#[0-9a-fA-F]{6}$").matches(raw)) return fallback
    return Color(android.graphics.Color.parseColor(raw))
}

fun Modifier.blockSurface(block: Block, isText: Boolean, scale: Float): Modifier {
    if (isText && (block.props["textSurfaceEnabled"] as? JsonPrimitive)?.content != "true") return this
    val width = appearanceNumber(block, "borderWidth", if (isText) 1f else 0f, 12f) * scale
    val shape = RoundedCornerShape((appearanceNumber(block, "borderRadius", if (isText) 12f else 0f, 999f) * scale).dp)
    return clip(shape)
        .background(appearanceColor(block, "backgroundColor", if (isText) Color.White else Color.Transparent))
        .border(width.dp, appearanceColor(block, "borderColor", Color(0xFFCBD5E1)), shape)
        // CSS borders consume interior space; Compose borders alone do not.
        .padding(width.dp)
}
